### PR TITLE
New version of FrameUtility

### PR DESCRIPTION
**Pull request by moi15moi**
The old version of FrameUtility did not work correctly. Now, it should always work.

This patch add 3 method in Convert.py (ms2frame, frame2ms and bound2frame) and a totally new version of FrameUtility.

Like you can see, at the line 81 of Convert.py, i add 0.011. I don't know why I have to do it, because I compared how pyonfx convert timestamp to ms (https://github.com/CoffeeStraw/PyonFX/blob/master/pyonfx/convert.py#L72) and aegisub (https://github.com/Ristellise/AegisubDC/blob/d4e6c9afef17953c2d62b874665a1bfb62949b32/libaegisub/ass/time.cpp#L29) and they gave the same ms. Maybe ffms2 don't index the frame the exact same way I do.

The bound2frame function is really usefull when you want to use FrameUtility with \t(). Like that, all the timing in the \t tag will be good

**Warning 1**:
I did not modify the **add function** of FrameUtility. It absolutely need to be modify by someone, because with the changes made, all the timing is now in frame (previously it was in ms), so there is no more fr.

**Warning 2**:
If you open a fake video with aegisub, this new FrameUtility will not work. I still don't know why, but aegisub does not index the fake video frame the same way that it index real video.

**Warning 3**:
This new version of FrameUtility still don't work with VFR (variable frame rate) video.

Example: `Convert.ms2frame(l.start_time, Fraction(24000/1001), "start")`

Thanks to Varde for his help!

### CODE REVIEW

**Comment by moi15moi** [comment]
Hum, finally, I think this line (https://github.com/moi15moi/PyonFX/blob/frameUtilityDebug/pyonfx/convert.py#L81) could be replace by this: image = math.ceil((float(ms + 0.5)/1000) * fps)

I find the 0.5 value here: https://github.com/Ristellise/AegisubDC/blob/master/libaegisub/common/vfr.cpp#L119

It seems to do the same job than what I have implemented. It would need more tests to be sure, because from what I understand from aegisub code, our video has a constant fps, so it should not go to the v1_parse method.

File: pyonfx/convert.py [hunk 1]
```diff
@@ -16,12 +16,15 @@
 # along with this program. If not, see http://www.gnu.org/licenses/.
 
 from __future__ import annotations
+from fractions import Fraction
 import re
 import math
 import colorsys
 from enum import Enum
+import string
```
Comment (CoffeeStraw) [inline pyonfx/convert.py:24]: Why is this import needed?

Comment (CoffeeStraw) [inline pyonfx/convert.py:24]: Why is this import needed?

File: pyonfx/utils.py [hunk 2]
```diff
@@ -163,32 +165,32 @@ class FrameUtility:
...
+                Convert.frame2ms(self.current_time, self.fps, "end"),
+                i + 1,
                 self.n,
             )
-            self.current_time += self.fr
-            self.fr_time += self.fr
 
-        # Compute values for the last frame, clamping the end_time of the frame at self.end_time
-        yield (round(self.current_time, 2), round(self.end_time, 2), self.n, self.n)
+            self.current_time = self.current_time + 1
```
Comment (CoffeeStraw) [inline pyonfx/utils.py:192]: Instead of changing this part, can't we just use the bound function at the beginning and then keep working with ms as it was before?

**Review by CoffeeStraw — CHANGES_REQUESTED** [review]
I've left you an initial comment for now that I think needs to be fixed first! After the main functionality is implemented I will contribute to fix documentation, code style and add tests!

File: pyonfx/convert.py (hunk 1 above)
Comment (JeremieBergeron) [reply pyonfx/convert.py:24]: We don't need it. I forget to delete it

Comment (moi15moi) [reply pyonfx/convert.py:24]: We don't need it. I forget to delete it.

File: pyonfx/utils.py (hunk 2 above)
Comment (moi15moi) [reply pyonfx/utils.py:192]: Which value will you use to make to get the next frame?

In frame, it is easy. If you have 1000. The next frame is 1001.

In ms, it is not that simple.

Comment (moi15moi) [reply pyonfx/utils.py:192]: It would need to be properly tested.

But, has frameutility "play" with frame, I think all calculations should be made in image

**Comment by CoffeeStraw** [comment]
Hey there, I've refactored your code and fixed FU.add.
It would be nice if you can further test it, I've played with it a little bit and it seems to work!
